### ISSUE TITLE
"needs-restarting -r" reports no reboot required if RTC is not in UTC

### ISSUE THREAD

**Issue by inknos** [rpm-software-management/dnf5#1184]
This issue comes from a ticket filled for dnf4
https://issues.redhat.com/browse/RHEL-14900

I am reporting two weird behaviours that I suspect are connected with a single issue.

This [PR](https://github.com/rpm-software-management/dnf-plugins-core/pull/518) fixes the first report partially, but beaker tests revealed that the wanted behavior is inconsistent with repeated testing.

### Reproducer 1
I could reproduce it with `dnf5-5.1.11-1.fc39.x86_64` on F39 with the steps:
1.
```bash
timedatectl set-local-rtc 1
```
2.
```bash
reboot
```
3.
```bash
dnf5 -y reinstall systemd
```
4.
```bash
dnf5 needs-restarting -r
...
Core libraries or services have been updated since boot-up:
  * systemd

Reboot is required to fully utilize these updates.
```
5.
```bash
reboot
```
6.
```bash
dnf5 needs-restarting -r
```

### Actual Results

```bash
Core libraries or services have been updated since boot-up:
  * systemd

Reboot is required to fully utilize these updates.
```

### Expected results
```
No core libraries or services have been updated since boot-up.
Reboot should not be necessary.
```

---

### Reproducer 2

1.
```bash
timedatectl set-local-rtc 1
timedatectl set-timezone Europe/Paris
```
2.
```bash
reboot
```
3.
```bash
dnf5 reinstall -y systemd
```
4.
```bash
dnf5 needs-restarting -r
```

### Expected results
```bash
Core libraries or services have been updated since boot-up:
  * systemd

Reboot is required to fully utilize these updates.
```

### Actual results
```
No core libraries or services have been updated since boot-up.
Reboot should not be necessary.
```

**Comment by evan-goode** [comment]
Related: https://bugzilla.redhat.com/show_bug.cgi?id=2237223

I can reproduce with your second reproducer. DNF 5 is getting an incorrect boot time from systemd:

/var/log/dnf5.log:
```
2024-01-22T14:36:31+0000 [3083] DEBUG Got boot time from systemd: 1705937754
```
1705937754 is Mon Jan 22 15:35:54 2024 UTC.

The `btime` in `/proc/stat` that you use in https://github.com/rpm-software-management/dnf-plugins-core/pull/518 seems to be reliable for me, but in containers it will have the host's boot time, not the container's.

Another possible fix is to get `UserspaceTimestampMonotonic` instead of `UserspaceTimestamp` and treat it like an uptime. This is a bit distinct from `/proc/uptime` since if we are in a container, systemd will either not be available (most typical case) or will report the correct uptime of the container, whereas `/proc/uptime` will have the uptime of the host.

Re: the tests, I encountered [this issue](https://github.com/rpm-software-management/ci-dnf-stack/pull/1403#issuecomment-1811311473) while enabling the CI tests for `dnf5 needs-restarting`, but I'm not sure whether that's relevant here.

**Comment by inknos** [comment]
> but in containers it will have the host's boot time, not the container's.

The command is heavily dependent on which virtual env you are in, which is a recurring issue for needs-restarting. We even considered wrapping `systemd-detect-virt` at some point

**Comment by inknos** [comment]
> Re: the tests, I encountered https://github.com/rpm-software-management/ci-dnf-stack/pull/1403#issuecomment-1811311473 while enabling the CI tests for dnf5 needs-restarting, but I'm not sure whether that's relevant here.

I am not sure this would be easily tested in our ci. currently, the only tests for this issue are run in beaker

**Comment by evan-goode** [comment]
> Another possible fix is to get UserspaceTimestampMonotonic instead of UserspaceTimestamp and treat it like an uptime. This is a bit distinct from /proc/uptime since if we are in a container, systemd will either not be available (most typical case) or will report the correct uptime of the container, whereas /proc/uptime will have the uptime of the host.

OK, I misunderstood what `UserspaceTimestampMonotonic` was. It's not an uptime at all: https://unix.stackexchange.com/a/453258, so that idea is out the window. But looking at `systemctl show`, I noticed that `UnitsLoadStartTimestamp` had the correct time even when `UserspaceTimestamp` was incorrect due to `set-local-rtc 1` after following your reproducer. So I prepared a PR that uses that timestamp instead: https://github.com/rpm-software-management/dnf5/pull/1198.

If `set-local-rtc 1` problems keep coming up, I'd rather point users to https://wiki.archlinux.org/title/System_time#UTC_in_Microsoft_Windows and say it's unsupported. `timedatectl` says so:

```
Warning: The system is configured to read the RTC time in the local time zone.
         This mode cannot be fully supported. It will create various problems
         with time zone changes and daylight saving time adjustments. The RTC
         time is never updated, it relies on external facilities to maintain it.
         If at all possible, use RTC in UTC by calling
         'timedatectl set-local-rtc 0'.
```

**Comment by evan-goode** [comment]
Should be fixed by #1198
